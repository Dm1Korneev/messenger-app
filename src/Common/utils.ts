
type Filter = Parameters<typeof Array.prototype.filter>[0]
export const distinct: Filter = (value, index, self) => self.indexOf(value) === index;
