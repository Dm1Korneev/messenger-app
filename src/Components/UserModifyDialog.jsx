import React from 'react';
import { ValidatorForm } from 'react-material-ui-form-validator';
import PropTypes from 'prop-types';

import withStyles from '@material-ui/core/styles/withStyles';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import FormHelperText from '@material-ui/core/FormHelperText';

import UserEmailField from 'Components/UserEmailField';
import UserNameField from 'Components/UserNameField';
import UserPasswordField from 'Components/UserPasswordField';
import AvatarSelector from 'Components/AvatarSelector';

class UserModifyDialog extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.user.name,
      email: props.user.email,
      password: '**********',
      nameIsModified: false,
      emailIsModified: false,
      avatarIsModified: false,
      passwordIsModified: false,
    };
    this.avatarFileInput = React.createRef();
  }

  avatarOnChange = () => {
    this.setState({
      avatarIsModified: true,
    });
  };

  handleInputChange = (event) => {
    const { target } = event;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const { name } = target;

    this.setState({
      [name]: value,
      [`${name}IsModified`]: true,
    });
  };

  onSubmit = () => {
    const {
      name,
      email,
      password,
      nameIsModified,
      emailIsModified,
      avatarIsModified,
      passwordIsModified,
    } = this.state;
    const { onSave, user } = this.props;

    let options = {};
    if (nameIsModified) {
      options = { ...options, name };
    }
    if (emailIsModified) {
      options = { ...options, email };
    }
    if (passwordIsModified) {
      options = { ...options, password };
    }
    if (avatarIsModified) {
      options = {
        ...options,
        avatar: this.avatarFileInput.current.files[0],
      };
    }

    onSave({ userId: user._id, options });
  };

  render() {
    const { closeUserModifyDialog, user, error } = this.props;
    const { name, email, password } = this.state;
    const { avatar } = user;

    return (
      <Dialog open onClose={closeUserModifyDialog}>
        <DialogTitle>Modify user</DialogTitle>
        <DialogContent>
          <ValidatorForm id="validatorForm" onSubmit={this.onSubmit}>
            {error && <FormHelperText error>{error}</FormHelperText>}
            <AvatarSelector
              onChange={this.avatarOnChange}
              avatar={avatar}
              avatarFileInput={this.avatarFileInput}
            />
            <UserNameField value={name} onChange={this.handleInputChange} />
            <UserEmailField value={email} onChange={this.handleInputChange} />
            <UserPasswordField
              value={password}
              onChange={this.handleInputChange}
            />
          </ValidatorForm>
        </DialogContent>
        <DialogActions>
          <Button type="submit" color="primary" form="validatorForm">
            Save
          </Button>
          <Button onClick={closeUserModifyDialog} color="primary">
            Close
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}

UserModifyDialog.defaultProps = {
  error: undefined,
};
UserModifyDialog.propTypes = {
  user: PropTypes.shape({
    _id: PropTypes.string.isRequired,
    avatar: PropTypes.string,
    name: PropTypes.string,
    email: PropTypes.string,
  }).isRequired,
  onSave: PropTypes.func.isRequired,
  closeUserModifyDialog: PropTypes.func.isRequired,
  error: PropTypes.string,
};

const styles = () => ({});

export default withStyles(styles)(UserModifyDialog);
