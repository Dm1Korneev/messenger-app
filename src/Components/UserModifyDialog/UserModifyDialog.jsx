import React, { useRef, useState } from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import FormHelperText from '@material-ui/core/FormHelperText';
import LinearProgress from '@material-ui/core/LinearProgress';
import { Field, Form, Formik } from 'formik';
import * as Yup from 'yup';

import UserEmailField from 'Components/UserEmailField';
import AvatarSelector from 'Components/AvatarSelector';
import UserNameField from 'Components/UserNameField';
import UserPasswordField from 'Components/UserPasswordField';
import {
  email as emailValidation,
  name as nameValidation,
  password as passwordValidation,
} from 'Common/validation';

const validationSchema = Yup.object().shape({
  email: emailValidation,
  password: passwordValidation,
  name: nameValidation,
});

const DEFAULT_PASSWORD = '**********';

const UserModifyDialog = ({
  user, error, closeUserModifyDialog, onSave,
}) => {
  const [avatarIsModified, setAvatarIsModified] = useState(false);

  const avatarFileInput = useRef();

  const avatarOnChange = () => {
    setAvatarIsModified(true);
  };

  const onSubmit = (values) => {
    const { name, email, password } = values;
    const options = { name, email };
    if (password !== DEFAULT_PASSWORD) {
      options.password = password;
    }
    if (avatarIsModified) {
      const [avatarValue] = avatarFileInput.current.files;
      options.avatar = avatarValue;
    }

    onSave({ userId: user._id, options });
  };

  return (
    <Dialog open onClose={closeUserModifyDialog}>
      <Formik
        initialValues={{
          name: user.name,
          email: user.email,
          password: DEFAULT_PASSWORD,
        }}
        validationSchema={validationSchema}
        onSubmit={onSubmit}
      >
        {({ isSubmitting, submitForm }) => (
          <Form>
            <DialogTitle>Modify user</DialogTitle>
            <DialogContent>
              {error && <FormHelperText error>{error}</FormHelperText>}
              <AvatarSelector
                onChange={avatarOnChange}
                avatar={user.avatar}
                avatarFileInput={avatarFileInput}
              />
              <Field component={UserNameField} name="name" />
              <Field component={UserEmailField} name="email" />
              <Field component={UserPasswordField} name="password" />
              {isSubmitting && <LinearProgress />}
            </DialogContent>
            <DialogActions>
              <Button
                type="submit"
                color="primary"
                form="validatorForm"
                disabled={isSubmitting}
                onClick={submitForm}
              >
                Save
              </Button>
              <Button onClick={closeUserModifyDialog} color="primary">
                Close
              </Button>
            </DialogActions>
          </Form>
        )}
      </Formik>
    </Dialog>
  );
};

UserModifyDialog.defaultProps = {
  error: undefined,
};
UserModifyDialog.propTypes = {
  user: PropTypes.shape({
    _id: PropTypes.string.isRequired,
    avatar: PropTypes.string,
    name: PropTypes.string,
    email: PropTypes.string,
  }).isRequired,
  onSave: PropTypes.func.isRequired,
  closeUserModifyDialog: PropTypes.func.isRequired,
  error: PropTypes.string,
};

export default UserModifyDialog;

