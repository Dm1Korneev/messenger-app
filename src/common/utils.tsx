export function distinct(value: any, index: number, self: Array<any>): boolean {
  return self.indexOf(value) === index;
}
