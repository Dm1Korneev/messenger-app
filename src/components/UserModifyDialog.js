import React from "react";
import { ValidatorForm } from "react-material-ui-form-validator";
import PropTypes from "prop-types";

// components
import AvatarSelector from "./AvatarSelector";
import UserPasswordField from "./UserPasswordField";
import UserNameField from "./UserNameField";
import UserEmailField from "./UserEmailField";

// material-ui
import withStyles from "@material-ui/core/styles/withStyles";
import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";
import FormHelperText from "@material-ui/core/FormHelperText";

class UserModifyDialog extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.user.name,
      email: props.user.email,
      password: "**********",
      nameIsModified: false,
      emailIsModified: false,
      avatarIsModified: false,
      passwordIsModified: false
    };
    this.avatarFileInput = React.createRef();
  }

  avatarOnChange = () => {
    this.setState({
      avatarIsModified: true
    });
  };

  handleInputChange = event => {
    const target = event.target;
    const value = target.type === "checkbox" ? target.checked : target.value;
    const name = target.name;

    this.setState({
      [name]: value,
      [name + "IsModified"]: true
    });
  };

  onSubmit = () => {
    const {
      name,
      email,
      password,
      nameIsModified,
      emailIsModified,
      avatarIsModified,
      passwordIsModified
    } = this.state;
    const { onSave, user } = this.props;

    let options = {};
    if (nameIsModified) {
      options = { ...options, name };
    }
    if (emailIsModified) {
      options = { ...options, email };
    }
    if (passwordIsModified) {
      options = { ...options, password };
    }
    if (avatarIsModified) {
      options = {
        ...options,
        avatar: this.avatarFileInput.current.files[0]
      };
    }

    onSave(user._id, options);
  };

  render() {
    const { closeUserModifyDialog, user, error } = this.props;
    const { name, email, password } = this.state;
    const { avatar } = user;

    return (
      <Dialog open onClose={closeUserModifyDialog}>
        <DialogTitle>Modify user</DialogTitle>
        <DialogContent>
          <ValidatorForm id="validatorForm" onSubmit={this.onSubmit}>
            {error && <FormHelperText error>{error}</FormHelperText>}
            <AvatarSelector
              onChange={this.avatarOnChange}
              avatar={avatar}
              avatarFileInput={this.avatarFileInput}
            />
            <UserNameField value={name} onChange={this.handleInputChange} />
            <UserEmailField value={email} onChange={this.handleInputChange} />
            <UserPasswordField
              value={password}
              onChange={this.handleInputChange}
            />
          </ValidatorForm>
        </DialogContent>
        <DialogActions>
          <Button type="submit" color="primary" form="validatorForm">
            Save
          </Button>
          <Button onClick={closeUserModifyDialog} color="primary">
            Close
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}

UserModifyDialog.propTypes = {
  user: PropTypes.shape({
    _id: PropTypes.string.isRequired,
    avatar: PropTypes.string,
    name: PropTypes.string,
    email: PropTypes.string
  }),
  onSave: PropTypes.func.isRequired,
  closeUserModifyDialog: PropTypes.func.isRequired,
  error: PropTypes.string
};

const styles = theme => ({});

export default withStyles(styles)(UserModifyDialog);
