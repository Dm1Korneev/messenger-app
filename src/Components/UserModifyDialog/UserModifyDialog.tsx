import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogTitle from '@mui/material/DialogTitle';
import FormHelperText from '@mui/material/FormHelperText';
import LinearProgress from '@mui/material/LinearProgress';
import { Field, Form, Formik } from 'formik';
import { useRef, useState } from 'react';
import * as Yup from 'yup';

import {
  email as emailValidation,
  name as nameValidation,
  password as passwordValidation,
} from 'Common/validation';
import { AvatarSelector } from 'Components/AvatarSelector';
import { UserEmailField } from 'Components/UserEmailField';
import { UserNameField } from 'Components/UserNameField';
import { UserPasswordField } from 'Components/UserPasswordField';
import { useUpdateUser, UseUpdateUserPayload, useCurrentUser } from 'Hooks';

const validationSchema = Yup.object().shape({
  email: emailValidation,
  password: passwordValidation,
  name: nameValidation,
});

type FromValues = {
  name: string;
  email: string;
  password: string;
}

const DEFAULT_PASSWORD = '**********';

type UserModifyDialogProps = {
  onClose: ()=>void
}

export const UserModifyDialog = ({ onClose }: UserModifyDialogProps) => {
  const { data: currentUser } = useCurrentUser();

  const { mutate: updateUser, error } = useUpdateUser();

  const [avatarIsModified, setAvatarIsModified] = useState(false);

  const avatarFileInput = useRef<HTMLInputElement>(null);

  if (!currentUser) {
    return null;
  }

  const avatarOnChange = () => {
    setAvatarIsModified(true);
  };

  const onSubmit = async ({ name, email, password }: FromValues) => {
    const files = avatarFileInput?.current?.files;
    const avatar = files ? files[0] : undefined;

    const modifyData: UseUpdateUserPayload = {
      name,
      email,
    };

    if (avatarIsModified) {
      modifyData.avatar = avatar;
    }
    if (password !== DEFAULT_PASSWORD && password) {
      modifyData.password = password;
    }

    updateUser({ userId: currentUser._id, modifyData }, { onSuccess: () => onClose() });
  };

  const initialValues: FromValues = {
    name: currentUser.name,
    email: currentUser.email,
    password: DEFAULT_PASSWORD,
  };

  return (
    <Dialog onClose={onClose} open>
      <Formik
        initialValues={initialValues}
        onSubmit={onSubmit}
        validationSchema={validationSchema}
      >
        {({ isSubmitting, submitForm }) => (
          <Form>
            <DialogTitle>Modify user</DialogTitle>
            <DialogContent>
              {error ? <FormHelperText error>{String(error)}</FormHelperText> : null}
              <AvatarSelector
                avatar={currentUser.avatar}
                avatarFileInput={avatarFileInput}
                onChange={avatarOnChange}
              />
              <Field component={UserNameField} name="name" />
              <Field component={UserEmailField} name="email" />
              <Field component={UserPasswordField} name="password" />
              {isSubmitting && <LinearProgress />}
            </DialogContent>
            <DialogActions>
              <Button
                color="primary"
                disabled={isSubmitting}
                onClick={submitForm}
                type="submit"
              >
                Save
              </Button>
              <Button color="primary" onClick={onClose}>
                Close
              </Button>
            </DialogActions>
          </Form>
        )}
      </Formik>
    </Dialog>
  );
};
