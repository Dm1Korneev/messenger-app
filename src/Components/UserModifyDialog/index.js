import commonHoc from 'Components/commonHoc';

import { modifyUser, setModifyUserDialogIsOpen } from 'Redux/actions';
import { MODIFY_USER } from 'Constants/actionNames';

import { currentUserSelector } from 'Selectors/session';
import { errorSelector } from 'Selectors/errors';

import UserModifyDialog from './UserModifyDialog';

const mapStateToProps = (state) => ({
  user: currentUserSelector(state),
  error: errorSelector(state, MODIFY_USER),
});

const mapDispatchToProps = {
  closeUserModifyDialog: () => setModifyUserDialogIsOpen(false),
  onSave: modifyUser,
};

const Container = commonHoc(UserModifyDialog, {
  mapStateToProps,
  mapDispatchToProps,
});

export { Container as UserModifyDialog };
