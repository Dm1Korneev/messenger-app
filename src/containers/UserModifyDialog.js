import { connect } from 'react-redux';

import { modifyUser, setUserModifyDialogIsOpen } from 'Redux/actions';
import UserModifyDialog from 'Components/UserModifyDialog';

const mapStateToProps = (state) => ({
  user: state.session.user,
  error: state.errors.MODIFY_USER,
});

const mapDispatchToProps = (dispatch) => ({
  closeUserModifyDialog: () => dispatch(setUserModifyDialogIsOpen(false)),
  onSave: (...args) => dispatch(modifyUser(...args)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(UserModifyDialog);
