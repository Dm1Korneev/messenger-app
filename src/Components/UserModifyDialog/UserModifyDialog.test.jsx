import React from 'react';
import { render } from '@testing-library/react';

import { mockComponent } from 'testing/utils';

import UserModifyDialog from './UserModifyDialog';

jest.mock('react-material-ui-form-validator', () => ({
  ValidatorForm: mockComponent('ValidatorForm'),
}));
jest.mock('@material-ui/core/Button', () => mockComponent('Button'));
jest.mock('@material-ui/core/Dialog', () => mockComponent('Dialog'));
jest.mock('@material-ui/core/DialogActions', () => mockComponent('DialogActions'));
jest.mock('@material-ui/core/DialogContent', () => mockComponent('DialogContent'));
jest.mock('@material-ui/core/DialogTitle', () => mockComponent('DialogTitle'));
jest.mock('@material-ui/core/FormHelperText', () => mockComponent('FormHelperText'));

jest.mock('Components/UserEmailField', () => mockComponent('UserEmailField'));
jest.mock('Components/UserNameField', () => mockComponent('UserNameField'));
jest.mock('Components/UserPasswordField', () => mockComponent('UserPasswordField'));
jest.mock('Components/AvatarSelector', () => mockComponent('AvatarSelector'));

const props = {
  user: {
    _id: '_id',
    avatar: 'avatar',
    name: 'name',
    email: 'email',
  },
  onSave: jest.fn(),
  closeUserModifyDialog: jest.fn(),
  error: 'error',
};

describe('render UserModifyDialog component', () => {
  const setup = () => render(
    <UserModifyDialog {...props} />,
  );

  test('snapshot test', async () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });
});

