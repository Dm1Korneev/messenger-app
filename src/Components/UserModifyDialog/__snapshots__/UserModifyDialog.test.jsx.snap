// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render UserModifyDialog component snapshot test 1`] = `
<DocumentFragment>
  <div
    data-originalcomponent="Dialog"
    test-prop-open="true"
  >
    <div
      data-originalcomponent="DialogTitle"
    >
      Modify user
    </div>
    <div
      data-originalcomponent="DialogContent"
    >
      <div
        data-originalcomponent="ValidatorForm"
        test-prop-id="\\"validatorForm\\""
      >
        <div
          data-originalcomponent="FormHelperText"
          test-prop-error="true"
        >
          error
        </div>
        <div
          data-originalcomponent="AvatarSelector"
          test-prop-avatar="\\"avatar\\""
          test-prop-avatarfileinput="{\\"current\\":null}"
        />
        <div
          data-originalcomponent="UserNameField"
          test-prop-value="\\"name\\""
        />
        <div
          data-originalcomponent="UserEmailField"
          test-prop-value="\\"email\\""
        />
        <div
          data-originalcomponent="UserPasswordField"
          test-prop-value="\\"**********\\""
        />
      </div>
    </div>
    <div
      data-originalcomponent="DialogActions"
    >
      <div
        data-originalcomponent="Button"
        test-prop-color="\\"primary\\""
        test-prop-form="\\"validatorForm\\""
        test-prop-type="\\"submit\\""
      >
        Save
      </div>
      <div
        data-originalcomponent="Button"
        test-prop-color="\\"primary\\""
      >
        Close
      </div>
    </div>
  </div>
</DocumentFragment>
`;
